File: src/features/companies/dto/update-company.dto.ts
import { Transform, Type } from 'class-transformer';
import { IsArray, IsEmail, IsNumber, IsOptional, IsString } from 'class-validator';

export class UpdateCompanyDto {
  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  name?: string;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  externalId?: string | null;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  website?: string | null;

  @IsOptional()
  @IsEmail()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim().toLowerCase() : value))
  mainEmail?: string | null;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  mainPhone?: string | null;

  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  companyTypeKeys?: string[];

  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  tagKeys?: string[];

  @IsOptional()
  @IsNumber()
  @Type(() => Number)
  rating?: number | null;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  notes?: string | null;
}

File: spring-boot-api/src/main/java/com/mytegroup/api/dto/companies/UpdateCompanyDto.java
package com.mytegroup.api.dto.companies;

import jakarta.validation.constraints.Email;

import java.util.List;

public record UpdateCompanyDto(
    String name,
    String externalId,
    String website,
    @Email(message = "Email must be valid")
    String mainEmail,
    String mainPhone,
    List<String> companyTypeKeys,
    List<String> tagKeys,
    Double rating,
    String notes
) {
    public UpdateCompanyDto {
        if (name != null) {
            name = name.trim();
        }
        if (externalId != null) {
            externalId = externalId.trim();
        }
        if (website != null) {
            website = website.trim();
        }
        if (mainEmail != null) {
            mainEmail = mainEmail.toLowerCase().trim();
        }
        if (mainPhone != null) {
            mainPhone = mainPhone.trim();
        }
        if (notes != null) {
            notes = notes.trim();
        }
    }
}

